### ISSUE TITLE
BaseNDArray.detach() can cause invalid shape information

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/nd4j#2614]
Reproduce using:
https://github.com/deeplearning4j/deeplearning4j/blob/6335b5e933f13d2bdb5c1f7e5896c45eb96839d7/deeplearning4j-core/src/test/java/org/deeplearning4j/nn/graph/graphnodes/TestGraphNodes.java#L542

Due to detach here:
https://github.com/deeplearning4j/deeplearning4j/blob/268a87e093ca81510fb95ab8bd08fe0a1ec34642/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java#L1808

Before detach:
![image](https://user-images.githubusercontent.com/2360237/36023397-6487bd42-0de0-11e8-9dc4-2c3576c749ad.png)

After detach:
![image](https://user-images.githubusercontent.com/2360237/36023410-6e382a0c-0de0-11e8-85be-7288c8745f1e.png)

Full stack trace:
```
java.lang.IndexOutOfBoundsException: 160

	at org.bytedeco.javacpp.indexer.Indexer.checkIndex(Indexer.java:90)
	at org.bytedeco.javacpp.indexer.DoubleRawIndexer.get(DoubleRawIndexer.java:59)
	at org.nd4j.linalg.api.buffer.BaseDataBuffer.getDouble(BaseDataBuffer.java:959)
	at org.nd4j.linalg.api.shape.Shape.getDouble(Shape.java:435)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.getDouble(BaseNDArray.java:1756)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.getDouble(BaseNDArray.java:3770)
	at org.nd4j.linalg.string.NDArrayStrings.format(NDArrayStrings.java:114)
	at org.nd4j.linalg.string.NDArrayStrings.format(NDArrayStrings.java:128)
	at org.nd4j.linalg.string.NDArrayStrings.format(NDArrayStrings.java:96)
	at org.nd4j.linalg.string.NDArrayStrings.format(NDArrayStrings.java:92)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.toString(BaseNDArray.java:4980)
	at org.deeplearning4j.nn.graph.graphnodes.TestGraphNodes.testLastTimeStepWithTransfer(TestGraphNodes.java:589)
```

I believe this started after the following PR was merged (but not 100% sure). It was about that time, however.
https://github.com/deeplearning4j/nd4j/pull/2600

**Comment by raver119** [comment]
Hypothetically fixed. Please confirm.

**Comment by AlexDBlack** [comment]
Confirmed fixed
